packages/ra-core/src/routing/useGetPathForRecord.ts

```diff
@@ -50,36 +51,36 @@ export const useGetPathForRecord = <RecordType extends RaRecord = RaRecord>(
         );
     }
     const getPathForRecord = useGetPathForRecordCallback<RecordType>(options);
+    const canAccess = useCanAccessCallback();
 
     // we initialize the path with the link value
-    const [path, setPath] = useState<string | false | undefined>(() => {
-        getPathForRecord({
-            record,
-            resource,
-            link,
-        }).then(resolvedLink => {
-            if (resolvedLink === false) {
-                // already set to false by default
-                return;
-            }
-            // update the path when the promise resolves
-            setPath(resolvedLink);
-        });
-
-        return false;
-    });
+    const [path, setPath] = useState<string | false | undefined>();
 
     // update the path if the record changes
     useEffect(() => {
-        getPathForRecord({
-            record,
-            resource,
-            link,
-        }).then(resolvedLink => {
+        const updatePath = async () => {
+            const resolvedLink = await getPathForRecord({
+                record,
+                resource,
+                link,
+            });
+            if (resolvedLink && ['edit', 'show'].includes(resolvedLink)) {
+                if (
+                    !(await canAccess({
+                        action: resolvedLink,
```

Comment:
`resolvedLink` is like `/posts/123`, it's not `edit` or `show`, am I right? If so, it's not what you want to use for the `canAccess` action.